### DISCUSSION TITLE
Why is postgresql@14 lib folder symlinked into /usr/local/lib?

### DISCUSSION THREAD

**Discussion by nickform-ost** [Homebrew/discussions#3584] [Tap maintenance and Homebrew development] [answered]
Related to #3579 , I'm wondering why the lib folder within the postgresql@14 folder in my Cellar (`/usr/local/Cellar/postgresql@14/14.5_1/lib/postgresql@14`) is symlinked as /usr/local/lib/postgresql@14? The system does not find the libraries within subdirectories of `/usr/local/lib` anyway so this must have some other purpose or be here by mistake.

One hypothesis I have about the purpose is that I can have a stable target for a symlink that is within `/usr/local/lib` by doing the following:
```
cd /usr/local/lib
ln -s postgresql@14/libpq.dylib libpq.dylib
```
If the actual package version (i.e. the `14.5_1` part) is later upgraded then I will not need to update the link created in the steps above. Is this why the folder is symlinked?

If yes, it seems like the openssl@1.1 formula could also benefit from this as we always have to update the symlinks when the final letter changes (e.g. `1.1.1o` --> `1.1.1q` as just happened).

**Comment by SMillerDev** [comment]
I don't have an answer about the choice for this directory but making random symlinks in Homebrew managed directories is just asking for trouble.

**Reply by nickform-ost** [reply]
The directory in which I suggested one might make a symlink (just for the purpose of explaining the hypothesis) is `/usr/local/lib` which is clearly not managed by Homebrew. I'd prefer not to have to make symlinks at all after installing brew packages and up to last week I didn't have to do so for the `postgresql` formula but having upgraded to the latest definition I now do. I'm just trying to figure out the recommended way to cope with that.

**Comment by Bo98** [comment, marked as answer]
I think the idea was to have multiple PostgreSQL versions (14, 13, 12, ...) available side-by-side (though this feature currently isn't available yet).

If you're linking against `libpq`, you probably want the `libpq` formula, which is much simpler. I think it's still keg-only like OpenSSL currently, so will involve `/usr/local/opt/libpq/lib`. I will investigate whether this is still necessary.

> If yes, it seems like the openssl@1.1 formula could also benefit from this as we always have to update the symlinks when the final letter changes (e.g. `1.1.1o` --> `1.1.1q` as just happened).

Use `/usr/local/opt/openssl@1.1/lib`

**Reply by nickform-ost** [reply]
Thanks for the suggestion about the `libpq` formula but I need postgresql@14 anyway and, since that comes with a (presumably version-correct) libpq, what is the recommended way of linking its copy into a location where it will be found by the system's default library resolution logic?

**Reply by nickform-ost** [reply]
Actually ignore that last question - let's try and stay focused on the original question. You say:

> I think the idea was to have multiple PostgreSQL versions (14, 13, 12, ...) available side-by-side (though this feature currently isn't available yet).

but how would the symlink I'm asking about help in any way towards this goal? On it's own it doesn't seem to have any valid purpose.

**Reply by carlocab** [reply]
`Cellar` paths are an implementation detail of `brew` and it's better to pretend they don't exist. Otherwise, you end up using paths that change on you at inopportune moments (which you seem to have already encountered).

Given that, the `/usr/local/lib/postgresql@14` symlink (or even the `/usr/local/opt/postgresql@14/lib/postgresql@14` directory) are the locations one is meant to look for the libraries that are installed by `postgresql@14`.

**Reply by nickform-ost** [reply]
Thanks @carlocab. I surmise from this that my hypothesis about the purpose of the symlink was broadly correct. This answers my question.

**Reply by carlocab** [reply]
I'd like to offer a caveat to that: the earlier suggestion that creating your own symlinks in `$(brew --prefix)/lib` is a recipe for pain in the long run. I would strongly recommend against doing that.

**Reply by nickform-ost** [reply]
Noted - thanks.

**Reply by nickform-ost** [reply]
Actually, can I ask you to point me to the recommended way to avoid the "recipe for pain" @carlocab so that I can make sure I'm not overlooking a better solution? Does it require setting `DYLD_FALLBACK_LIBRARY_PATH` or something similar?

**Reply by nickform-ost** [reply]
A little bump @carlocab. I want to be respectful of your time and the fact that you contribute it to this open source software for the benefit of everyone. However, I really think a link to the recommended way would completely close this issue off and hopefully be useful to future visitors so if you can find a moment and have such a link, it would be very much appreciated. Thanks.

**Reply by carlocab** [reply]
Hey, sorry, I get way too many GitHub notifications so I don't get to read all of them.

The recommended way really depends on what you're trying to do, exactly. But the general idea is that software should not assume that `libpq` (or any other user-installed library, for that matter) is not in the default `dyld` search path. In the case of `libpq`, this may require querying `pg_config --libdir`, or some other mechanism for querying the install location of `libpq`.

`DYLD_FALLBACK_LIBRARY_PATH` may work for some things, but not others. Apple doesn't really like it either:
```
❯ FOO=foo /bin/bash -c 'echo ${FOO-empty}'
foo
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
empty
❯ FOO=foo /usr/bin/python3 -c 'import os; print(os.getenv("FOO"))'
foo
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/bin/python3 -c 'import os; print(os.getenv("DYLD_FALLBACK_LIBRARY_PATH"))'
None
```
The only way to avoid the behaviour above is to disable SIP, or to only use binaries that are not protected by SIP:
```
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/local/bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
/usr/lib
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/bin/env /usr/local/bin/bash -c 'echo ${DYLD_FALLBACK_LIBRARY_PATH-empty}'
empty
❯ DYLD_FALLBACK_LIBRARY_PATH=/usr/lib /usr/local/bin/python3 -c 'import os; print(os.getenv("DYLD_FALLBACK_LIBRARY_PATH"))'
/usr/lib
```

**Reply by nickform-ost** [reply]
No need to apologise at all. Thank you very much for coming back and adding that explanation. 

Just in case any future reader is wondering: I gather, from Googling, that SIP is macOS's System Integrity Protection.
